### DISCUSSION TITLE
How to correctly override `BinaryIO.write`?

### DISCUSSION THREAD

**Discussion by rpmrmartin** [microsoft/pyright#5275] [Q&A] [answered]
I have written a `BinaryIO` decorator, but I'm having trouble adding type hints to the `write` method that will pass Pyright without using `Any`. Here's my first attempt:

```python
# pyright: strict

from typing import BinaryIO

class Decorator(BinaryIO):

	def __init__(self, file: BinaryIO) -> None:
		self._file = file
    
	def write(self, s: bytes) -> int:
		return self._file.write(s)
```

This passed Pyright 1.1.311, but fails 1.1.313 with the following error:
```
/app/temp.py
  /app/temp.py:10:6 - error: Method "write" overrides class "IO" in an incompatible manner
    Parameter 2 type mismatch: base parameter is type "str", override parameter is type "bytes"
    Return type mismatch: base method returns type "int", override returns type "None"
      "str" is incompatible with "bytes"
      Type "None" cannot be assigned to type "int" (reportIncompatibleMethodOverride)
1 error, 0 warnings, 0 informations
```

Looking at typeshed, I think Pyright is complaining about the method not being compatible with the following overload:
```python
@overload
def write(self: IO[str], __s: str) -> int: ...
```

But `self` is not an `IO[str]`, so I don't think that overload applies. Is this a bug or do I misunderstand?

I also tried copying the annotations from typeshed as closely as I could. I checked this with Pyright built from d20a39cef353263a5d63a3924fa1d58dff4ff26f so it wouldn't complain about importing `Buffer`:

```python
# pyright: strict

from typing import Any, AnyStr, BinaryIO, overload

from typing_extensions import Buffer

class Decorator(BinaryIO):

	def __init__(self, file: BinaryIO) -> None:
		self._file = file

	@overload
	def write(self, s: str) -> int: ...

	@overload
	def write(self, s: Buffer) -> int: ...

	@overload
	def write(self, s: AnyStr) -> int: ...
	
	def write(self, s: Any) -> int:
		return self._file.write(s)
```

For this, Pyright generates the following error:
```
/app/temp.py
  /app/temp.py:21:6 - error: Method "write" overrides class "IO" in an incompatible manner
    Overloads for override method must be in the same order as the base method (reportIncompatibleMethodOverride)
1 error, 0 warnings, 0 informations
```

The overloads are in the same order as the base method, so I don't understand what Pyright is trying to tell me here, unless the overloads do not match the base method closely enough. If this is not a bug, could someone explain why Pyright is generating this error?

**Comment by erictraut** [comment, marked as answer]
There is a bug in the logic for overloaded overrides, which I added recently. I'll fix this in the next release.

That said, to avoid the `reportIncompatibleMethodOverride` error, you'll need to copy the overload signatures from typeshed exactly. That includes the use of a positional-only parameter `__s` rather than a named parameter `s`. It also includes type annotations for the `self` parameter on the first two overloads.

```python
class Decorator(BinaryIO):
    def __init__(self, file: BinaryIO) -> None:
        self._file = file

    @overload
    def write(self: IO[str], s: str, /) -> int:
        ...

    @overload
    def write(self: IO[bytes], s: ReadableBuffer, /) -> int:
        ...

    @overload
    def write(self, s: AnyStr, /) -> int:
        ...

    def write(self, s: Any, /) -> int:
        return self._file.write(s)
```

**Reply by rpmrmartin** [reply]
Thank you for your help. I cannot copy the overload signatures exactly because `ReadableBuffer` comes from an internal `_typeshed` package  and typing_extensions has no equivalent. `ReadableBuffer` is an alias of `Buffer`, but using `Buffer` in its place is giving me the same `reportIncompatibleMethodOverride` error. Do you have any suggestions on how to overcome this?

**Reply by erictraut** [reply]
You can use `Buffer` in place of `ReadableBuffer`. You won't be able to eliminate the `reportIncompatibleMethodOverride` until the next release of pyright, which addresses the bug. I publish a new version of pyright weekly on Tuesdays, so you won't have to wait long.

**Reply by rpmrmartin** [reply]
Thanks again!
